core/src/main/java/org/infinispan/marshall/core/internal/ExternalJBossMarshaller.java

```diff
@@ -0,0 +1,226 @@
+package org.infinispan.marshall.core.internal;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.OutputStream;
+
+import org.infinispan.commons.io.ByteBuffer;
+import org.infinispan.commons.marshall.BufferSizePredictor;
+import org.infinispan.commons.marshall.Externalizer;
+import org.infinispan.commons.marshall.MarshallableTypeHints;
+import org.infinispan.commons.marshall.StreamingMarshaller;
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.marshall.core.JBossMarshaller;
+import org.jboss.marshalling.ByteInput;
+import org.jboss.marshalling.ByteOutput;
+import org.jboss.marshalling.ObjectTable;
+import org.jboss.marshalling.Unmarshaller;
+
+final class ExternalJBossMarshaller implements StreamingMarshaller {
+
+   final MarshallableTypeHints marshallableTypeHints = new MarshallableTypeHints();
+   final JBossMarshaller marshaller;
```

Comment:
The distinction between internal and external marshallers seems a bit blurry here...

IMO the two should never mix. The user-supplied marshaller should be used for all keys and values, and never for internal objects, and the internal marshaller should not be used to write internal objects that happen to be contained in user objects. If some internal objects really need to work as part of user objects, we should make them `Serializable` and/or we should allow the user to call our internal marshaller explicitly, but the format of user objects in a cache store should never change just because we added a new internal externalizer for a JDK class.